File: android/src/main/kotlin/com/gtbluesky/fusion/engine/FusionEngineBinding.kt
package com.gtbluesky.fusion.engine

import com.gtbluesky.fusion.Fusion
import com.gtbluesky.fusion.constant.FusionConstant
import com.gtbluesky.fusion.container.FusionStackManager
import com.gtbluesky.fusion.event.FusionEventManager
import com.gtbluesky.fusion.navigator.FusionNavigator
import com.gtbluesky.fusion.navigator.FusionRouteType
import com.gtbluesky.fusion.event.FusionEventType
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.systemchannels.PlatformChannel
import io.flutter.embedding.engine.systemchannels.PlatformViewsChannel
import io.flutter.plugin.common.BasicMessageChannel
import io.flutter.plugin.common.StandardMessageCodec

internal class FusionEngineBinding(engine: FlutterEngine?) {
    private var hostPush: BasicMessageChannel<Any>? = null
    private var hostDestroy: BasicMessageChannel<Any>? = null
    private var hostRestore: BasicMessageChannel<Any>? = null
    private var hostSync: BasicMessageChannel<Any>? = null
    private var hostDispatchEvent: BasicMessageChannel<Any>? = null
    private var flutterCreate: BasicMessageChannel<Any>? = null
    private var flutterSwitchTop: BasicMessageChannel<Any>? = null
    private var flutterRestore: BasicMessageChannel<Any>? = null
    private var flutterDestroy: BasicMessageChannel<Any>? = null
    private var flutterPush: BasicMessageChannel<Any>? = null
    private var flutterReplace: BasicMessageChannel<Any>? = null
    private var flutterPop: BasicMessageChannel<Any>? = null
    private var flutterMaybePop: BasicMessageChannel<Any>? = null
    private var flutterRemove: BasicMessageChannel<Any>? = null
    private var flutterNotifyPageVisible: BasicMessageChannel<Any>? = null
    private var flutterNotifyPageInvisible: BasicMessageChannel<Any>? = null
    private var flutterNotifyEnterForeground: BasicMessageChannel<Any>? = null
    private var flutterNotifyEnterBackground: BasicMessageChannel<Any>? = null
    private var flutterDispatchEvent: BasicMessageChannel<Any>? = null
    private var flutterCheckStyle: BasicMessageChannel<Any>? = null

    var engine: FlutterEngine? = null
        private set

    private val historyList: List<Map<String, Any?>>
        get() {
            return FusionStackManager.containerStack.map {
                mapOf(
                    "uniqueId" to it.get()?.uniqueId(),
                    "history" to it.get()?.history()
                )
            }
        }

    init {
        this.engine = engine
        engine?.let {
            val binaryMessenger = it.dartExecutor.binaryMessenger
            val messageCodec = StandardMessageCodec.INSTANCE
            hostPush = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/host/push",
                messageCodec
            )
            hostDestroy = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/host/destroy",
                messageCodec
            )
            hostRestore = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/host/restore",
                messageCodec
            )
            hostSync = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/host/sync",
                messageCodec
            )
            hostDispatchEvent = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/host/dispatchEvent",
                messageCodec
            )
            flutterCreate = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/create",
                messageCodec
            )
            flutterSwitchTop = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/switchTop",
                messageCodec
            )
            flutterRestore = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/restore",
                messageCodec
            )
            flutterDestroy = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/destroy",
                messageCodec
            )
            flutterPush = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/push",
                messageCodec
            )
            flutterReplace = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/replace",
                messageCodec
            )
            flutterPop = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/pop",
                messageCodec
            )
            flutterMaybePop = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/maybePop",
                messageCodec
            )
            flutterRemove = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/remove",
                messageCodec
            )
            flutterNotifyPageVisible = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/notifyPageVisible",
                messageCodec
            )
            flutterNotifyPageInvisible = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/notifyPageInvisible",
                messageCodec
            )
            flutterNotifyEnterForeground = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/notifyEnterForeground",
                messageCodec
            )
            flutterNotifyEnterBackground = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/notifyEnterBackground",
                messageCodec
            )
            flutterDispatchEvent = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/dispatchEvent",
                messageCodec
            )
            flutterCheckStyle = BasicMessageChannel(
                binaryMessenger,
                "${FusionConstant.FUSION_CHANNEL}/flutter/checkStyle",
                messageCodec
            )
        }
    }

    @Suppress("UNCHECKED_CAST")
    fun attach() {
        hostPush?.setMessageHandler { message, reply ->
            if (message !is Map<*, *>) {
                reply.reply(null)
                return@setMessageHandler
            }
            val name = message["name"] as? String
            if (name == null) {
                reply.reply(null)
                return@setMessageHandler
            }
            val args = message["args"] as? Map<String, Any>
            val type = message["type"] as? Int
            if (type == FusionRouteType.FLUTTER_WITH_CONTAINER.ordinal) {
                Fusion.delegate.pushFlutterRoute(name, args)
            } else if (type == FusionRouteType.NATIVE.ordinal) {
                Fusion.delegate.pushNativeRoute(name, args)
            }
            reply.reply(null)
        }
        hostDestroy?.setMessageHandler { message, reply ->
            if (message !is Map<*, *>) {
                reply.reply(null)
                return@setMessageHandler
            }
            val uniqueId = message["uniqueId"] as? String
            if (uniqueId == null) {
                reply.reply(false)
                return@setMessageHandler
            }
            val container = FusionStackManager.findContainer(uniqueId)
            if (container != null) {
                FusionStackManager.closeContainer(container)
                reply.reply(true)
            } else {
                reply.reply(false)
            }
        }
        hostRestore?.setMessageHandler { _, reply ->
            reply.reply(historyList)
        }
        hostSync?.setMessageHandler { message, reply ->
            if (message !is Map<*, *>) {
                reply.reply(null)
                return@setMessageHandler
            }
            val uniqueId = message["uniqueId"] as? String
            val history = message["history"] as? List<Map<String, Any?>>
            if (uniqueId == null || history == null) {
                reply.reply(false)
                return@setMessageHandler
            }
            FusionStackManager.findContainer(uniqueId)?.history()?.let {
                it.clear()
                it.addAll(history)
            }
            reply.reply(true)
        }
        hostDispatchEvent?.setMessageHandler { message, reply ->
            if (message !is Map<*, *>) {
                reply.reply(null)
                return@setMessageHandler
            }
            val name = message["name"] as? String
            if (name == null) {
                reply.reply(null)
                return@setMessageHandler
            }
            val args = message["args"] as? Map<String, Any>
            FusionEventManager.send(name, args, FusionEventType.NATIVE)
            reply.reply(null)
        }
    }

    // external function
    fun push(name: String, args: Map<String, Any>?, type: FusionRouteType) {
        flutterPush?.send(
            mapOf(
                "name" to name,
                "args" to args,
                "type" to type.ordinal
            )
        )
    }

    fun replace(name: String, args: Map<String, Any>?) {
        flutterReplace?.send(
            mapOf(
                "name" to name,
                "args" to args
            )
        )
    }

    fun pop(result: Any?) {
        flutterPop?.send(
            mapOf(
                "result" to result
            )
        )
    }

    fun maybePop(result: Any?) {
        flutterMaybePop?.send(
            mapOf(
                "result" to result
            )
        )
    }

    fun remove(name: String) {
        flutterRemove?.send(
            mapOf(
                "name" to name,
            )
        )
    }

    // internal function
    fun create(uniqueId: String, name: String, args: Map<String, Any>? = null) {
        flutterCreate?.send(
            mapOf(
                "uniqueId" to uniqueId,
                "name" to name,
                "args" to args
            )
        )
    }

    fun switchTop(uniqueId: String, callback: () -> Unit) {
        flutterSwitchTop?.send(
            mapOf(
                "uniqueId" to uniqueId,
            )
        ) {
            callback()
        }
    }

    /**
     * Restore the specified container in flutter side
     * @param uniqueId: container's uniqueId
     * @param history: container's history
     */
    fun restore(uniqueId: String, history: List<Map<String, Any?>>) {
        flutterRestore?.send(
            mapOf(
                "uniqueId" to uniqueId,
                "history" to history,
            )
        )
    }

    /**
     * Destroy the specified container in flutter side
     * @param uniqueId: container's uniqueId
     */
    fun destroy(uniqueId: String) {
        if (!FusionStackManager.isAttached()) {
            engine?.let {
                // it.platformViewsController.attach(null, null, it.dartExecutor)
                try {
                    val platformViewsChannelField =
                        it.platformViewsController.javaClass.getDeclaredField("platformViewsChannel")
                    platformViewsChannelField.isAccessible = true
                    val platformViewsChannel = PlatformViewsChannel(it.dartExecutor)
                    platformViewsChannelField.set(it.platformViewsController, platformViewsChannel)
                    val channelHandlerField =
                        it.platformViewsController.javaClass.getDeclaredField("channelHandler")
                    channelHandlerField.isAccessible = true
                    val channelHandler =
                        channelHandlerField.get(it.platformViewsController) as? PlatformViewsChannel.PlatformViewsHandler
                    platformViewsChannel.setPlatformViewsHandler(channelHandler)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
        flutterDestroy?.send(
            mapOf(
                "uniqueId" to uniqueId,
            )
        )
    }

    fun notifyPageVisible(uniqueId: String) {
        flutterNotifyPageVisible?.send(
            mapOf(
                "uniqueId" to uniqueId,
            )
        )
    }

    fun notifyPageInvisible(uniqueId: String) {
        flutterNotifyPageInvisible?.send(
            mapOf(
                "uniqueId" to uniqueId,
            )
        )
    }

    fun notifyEnterForeground() {
        flutterNotifyEnterForeground?.send(null)
    }

    fun notifyEnterBackground() {
        flutterNotifyEnterBackground?.send(null)
    }

    fun dispatchEvent(name: String, args: Map<String, Any>?) {
        val msg = mapOf("name" to name, "args" to args)
        flutterDispatchEvent?.send(msg)
    }

    @Suppress("UNCHECKED_CAST")
    fun checkStyle(callback: (systemChromeStyle: PlatformChannel.SystemChromeStyle) -> Unit) {
        flutterCheckStyle?.send(null) {
            val systemChromeStyle = decodeSystemChromeStyle(it as? Map<String, Any>)
                ?: return@send
            callback(systemChromeStyle)
        }
    }

    private fun decodeSystemChromeStyle(styleMap: Map<String, Any>?): PlatformChannel.SystemChromeStyle? {
        if (styleMap == null) {
            return null
        }
        val statusBarColor = styleMap["statusBarColor"] as? Int
        val statusBarIconBrightness =
            if (styleMap["statusBarIconBrightness"] == "Brightness.light") {
                PlatformChannel.Brightness.LIGHT
            } else {
                PlatformChannel.Brightness.DARK
            }
        val systemStatusBarContrastEnforced =
            styleMap["systemStatusBarContrastEnforced"] as? Boolean
        val systemNavigationBarColor = styleMap["systemNavigationBarColor"] as? Int
        val systemNavigationBarIconBrightness =
            if (styleMap["systemNavigationBarIconBrightness"] == "Brightness.light") {
                PlatformChannel.Brightness.LIGHT
            } else {
                PlatformChannel.Brightness.DARK
            }
        val systemNavigationBarDividerColor = styleMap["systemNavigationBarDividerColor"] as? Int
        val systemNavigationBarContrastEnforced =
            styleMap["systemNavigationBarContrastEnforced"] as? Boolean
        return PlatformChannel.SystemChromeStyle(
            statusBarColor,
            statusBarIconBrightness,
            systemStatusBarContrastEnforced,
            systemNavigationBarColor,
            systemNavigationBarIconBrightness,
            systemNavigationBarDividerColor,
            systemNavigationBarContrastEnforced
        )
    }

    fun detach() {
        hostPush?.setMessageHandler(null)
        hostPush = null
        hostDestroy?.setMessageHandler(null)
        hostDestroy = null
        hostRestore?.setMessageHandler(null)
        hostRestore = null
        hostSync?.setMessageHandler(null)
        hostSync = null
        hostDispatchEvent?.setMessageHandler(null)
        hostDispatchEvent = null
        flutterCreate = null
        flutterSwitchTop = null
        flutterRestore = null
        flutterDestroy = null
        flutterPush = null
        flutterReplace = null
        flutterPop = null
        flutterMaybePop = null
        flutterRemove = null
        flutterNotifyPageVisible = null
        flutterNotifyPageInvisible = null
        flutterNotifyEnterForeground = null
        flutterNotifyEnterBackground = null
        flutterDispatchEvent = null
        flutterCheckStyle = null
        engine?.destroy()
        engine = null
    }
}

File: ios/Classes/Engine/FusionEngineBinding.swift
//
//  FusionEngineBinding.swift
//  fusion
//
//  Created by gtbluesky on 2022/3/10.
//

import Flutter
import Foundation

internal class FusionEngineBinding: NSObject {
    private var hostPush: FlutterBasicMessageChannel? = nil
    private var hostDestroy: FlutterBasicMessageChannel? = nil
    private var hostRestore: FlutterBasicMessageChannel? = nil
    private var hostSync: FlutterBasicMessageChannel? = nil
    private var hostDispatchEvent: FlutterBasicMessageChannel? = nil
    private var hostRemoveMaskView: FlutterBasicMessageChannel? = nil
    private var flutterCreate: FlutterBasicMessageChannel? = nil
    private var flutterSwitchTop: FlutterBasicMessageChannel? = nil
    private var flutterRestore: FlutterBasicMessageChannel? = nil
    private var flutterDestroy: FlutterBasicMessageChannel? = nil
    private var flutterPush: FlutterBasicMessageChannel? = nil
    private var flutterReplace: FlutterBasicMessageChannel? = nil
    private var flutterPop: FlutterBasicMessageChannel? = nil
    private var flutterMaybePop: FlutterBasicMessageChannel? = nil
    private var flutterRemove: FlutterBasicMessageChannel? = nil
    private var flutterNotifyPageVisible: FlutterBasicMessageChannel? = nil
    private var flutterNotifyPageInvisible: FlutterBasicMessageChannel? = nil
    private var flutterNotifyEnterForeground: FlutterBasicMessageChannel? = nil
    private var flutterNotifyEnterBackground: FlutterBasicMessageChannel? = nil
    private var flutterDispatchEvent: FlutterBasicMessageChannel? = nil
    private var flutterCheckStyle: FlutterBasicMessageChannel? = nil

    var engine: FlutterEngine? = nil

    private var historyList: [Dictionary<String, Any?>] {
        get {
            FusionStackManager.instance.containerStack.map {
                [
                    "uniqueId": $0.value?.uniqueId,
                    "history": $0.value?.history
                ]
            }
        }
    }

    init(_ engine: FlutterEngine?) {
        super.init()
        self.engine = engine;
        guard let engine = engine else {
            return
        }
        let binaryMessenger = engine.binaryMessenger
        hostPush = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/host/push", binaryMessenger: binaryMessenger)
        hostDestroy = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/host/destroy", binaryMessenger: binaryMessenger)
        hostRestore = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/host/restore", binaryMessenger: binaryMessenger)
        hostSync = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/host/sync", binaryMessenger: binaryMessenger)
        hostDispatchEvent = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/host/dispatchEvent", binaryMessenger: binaryMessenger)
        hostRemoveMaskView = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/host/removeMaskView", binaryMessenger: binaryMessenger)
        flutterCreate = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/create", binaryMessenger: binaryMessenger)
        flutterSwitchTop = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/switchTop", binaryMessenger: binaryMessenger)
        flutterRestore = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/restore", binaryMessenger: binaryMessenger)
        flutterDestroy = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/destroy", binaryMessenger: binaryMessenger)
        flutterPush = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/push", binaryMessenger: binaryMessenger)
        flutterReplace = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/replace", binaryMessenger: binaryMessenger)
        flutterPop = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/pop", binaryMessenger: binaryMessenger)
        flutterMaybePop = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/maybePop", binaryMessenger: binaryMessenger)
        flutterRemove = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/remove", binaryMessenger: binaryMessenger)
        flutterNotifyPageVisible = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/notifyPageVisible", binaryMessenger: binaryMessenger)
        flutterNotifyPageInvisible = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/notifyPageInvisible", binaryMessenger: binaryMessenger)
        flutterNotifyEnterForeground = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/notifyEnterForeground", binaryMessenger: binaryMessenger)
        flutterNotifyEnterBackground = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/notifyEnterBackground", binaryMessenger: binaryMessenger)
        flutterDispatchEvent = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/dispatchEvent", binaryMessenger: binaryMessenger)
        flutterCheckStyle = FlutterBasicMessageChannel(name: "\(FusionConstant.FUSION_CHANNEL)/flutter/checkStyle", binaryMessenger: binaryMessenger)
    }

    func attach() {
        hostPush?.setMessageHandler { (message: Any?, reply: @escaping FlutterReply) in
            guard let dict = message as? Dictionary<String, Any>, let name = dict["name"] as? String else {
                reply(nil)
                return
            }
            let args = dict["args"] as? Dictionary<String, Any>
            let type = dict["type"] as? Int
            if type == FusionRouteType.flutterWithContainer.rawValue {
                Fusion.instance.delegate?.pushFlutterRoute(name: name, args: args)
            } else if type == FusionRouteType.native.rawValue {
                Fusion.instance.delegate?.pushNativeRoute(name: name, args: args)
            }
            reply(nil)
        }
        hostDestroy?.setMessageHandler { (message: Any?, reply: @escaping FlutterReply) in
            guard let dict = message as? Dictionary<String, Any>, let uniqueId = dict["uniqueId"] as? String else {
                reply(false)
                return
            }
            if let container = FusionStackManager.instance.findContainer(uniqueId) {
                FusionStackManager.instance.closeContainer(container)
                reply(true)
            } else {
                reply(false)
            }
        }
        hostRestore?.setMessageHandler { (message: Any?, reply: @escaping FlutterReply) in
            reply(self.historyList)
        }
        hostSync?.setMessageHandler { (message: Any?, reply: @escaping FlutterReply) in
            guard let dict = message as? Dictionary<String, Any>, let hostPopGesture = dict["hostPopGesture"] as? Bool, let uniqueId = dict["uniqueId"] as? String, let history = dict["history"] as? [Dictionary<String, Any?>] else {
                reply(false)
                return
            }
            let container = FusionStackManager.instance.findContainer(uniqueId)
            container?.history.removeAll()
            container?.history.append(contentsOf: history)
            if container?.history.count == 1 && hostPopGesture {
                self.enablePopGesture()
            } else {
                self.disablePopGesture()
            }
            reply(true)
        }
        hostDispatchEvent?.setMessageHandler { (message: Any?, reply: @escaping FlutterReply) in
            guard let dict = message as? Dictionary<String, Any>, let name = dict["name"] as? String else {
                reply(nil)
                return
            }
            let args = dict["args"] as? Dictionary<String, Any>
            FusionEventManager.instance.send(name, args: args, type: .native)
            reply(nil)
        }
        hostRemoveMaskView?.setMessageHandler { (message: Any?, reply: @escaping FlutterReply) in
            guard let dict = message as? Dictionary<String, Any>, let uniqueId = dict["uniqueId"] as? String else {
                reply(nil)
                return
            }
            FusionStackManager.instance.findContainer(uniqueId)?.removeMask()
            reply(nil)
        }
    }

    // external function
    func push(_ name: String, args: Dictionary<String, Any>?, type: FusionRouteType) {
        flutterPush?.sendMessage([
            "name": name,
            "args": args as Any,
            "type": type.rawValue
        ])
    }

    func replace(_ name: String, args: Dictionary<String, Any>?) {
        flutterReplace?.sendMessage([
            "name": name,
            "args": args as Any
        ])
    }

    func pop(_ result: Any?) {
        flutterPop?.sendMessage([
            "result": result
        ])
    }

    func maybePop(_ result: Any?) {
        flutterMaybePop?.sendMessage([
            "result": result
        ])
    }

    func remove(_ name: String) {
        flutterRemove?.sendMessage([
            "name": name,
        ])
    }

    // internal function
    func enablePopGesture() {
        (UIApplication.roofViewController as? FusionPopGestureHandler)?.enablePopGesture()
    }

    func disablePopGesture() {
        (UIApplication.roofViewController as? FusionPopGestureHandler)?.disablePopGesture()
    }

    func create(_ uniqueId: String, name: String, args: Dictionary<String, Any>?) {
        flutterCreate?.sendMessage([
            "uniqueId": uniqueId,
            "name": name,
            "args": args as Any
        ])
    }

    func switchTop(_ uniqueId: String, _ callback: @escaping () -> Void) {
        flutterSwitchTop?.sendMessage(
            ["uniqueId": uniqueId],
            reply: { (_) in
                callback()
            }
        )
    }

    /**
     Restore the specified container in flutter side
     - Parameters:
       - uniqueId: container's uniqueId
       - history: container's history
     */
    func restore(_ uniqueId: String, history: [Dictionary<String, Any?>]) {
        flutterRestore?.sendMessage([
            "uniqueId": uniqueId,
            "history": history
        ])
    }

    /**
     Destroy the specified container in flutter side
     - Parameter uniqueId: container's uniqueId
     */
    func destroy(_ uniqueId: String) {
        flutterDestroy?.sendMessage([
            "uniqueId": uniqueId
        ])
    }

    func notifyPageVisible(_ uniqueId: String) {
        flutterNotifyPageVisible?.sendMessage([
            "uniqueId": uniqueId
        ])
    }

    func notifyPageInvisible(_ uniqueId: String) {
        flutterNotifyPageInvisible?.sendMessage([
            "uniqueId": uniqueId
        ])
    }

    func notifyEnterForeground() {
        flutterNotifyEnterForeground?.sendMessage(nil)
    }

    func notifyEnterBackground() {
        flutterNotifyEnterBackground?.sendMessage(nil)
    }

    func dispatchEvent(_ name: String, _ args: Dictionary<String, Any>?) {
        let msg: Dictionary<String, Any?> = ["name": name, "args": args]
        flutterDispatchEvent?.sendMessage(msg)
    }

    func checkStyle(_ callback: @escaping (_ statusBarStyle: UIStatusBarStyle) -> Void) {
        let infoValue = Bundle.main.object(forInfoDictionaryKey: "UIViewControllerBasedStatusBarAppearance") as? Bool
        if infoValue == false {
            return
        }
        flutterCheckStyle?.sendMessage(nil, reply: { (reply) in
            if (reply == nil) {
                return
            }
            if (reply is FlutterError) {
                return
            }
            if reply as? NSObject == FlutterMethodNotImplemented {
                return
            }
            var statusBarStyle: UIStatusBarStyle?
            if let map = reply as? Dictionary<String, Any> {
                let brightness = map["statusBarBrightness"] as? String
                if brightness == nil {
                    return
                }
                if brightness == "Brightness.dark" {
                    statusBarStyle = .lightContent
                } else if brightness == "Brightness.light" {
                    if #available(iOS 13, *) {
                        statusBarStyle = .darkContent
                    } else {
                        statusBarStyle = .default
                    }
                } else {
                    return
                }
            }
            guard let statusBarStyle = statusBarStyle else {
                return
            }
            callback(statusBarStyle)
        })
    }

    func detach() {
        hostPush?.setMessageHandler(nil)
        hostPush = nil
        hostDestroy?.setMessageHandler(nil)
        hostDestroy = nil
        hostRestore?.setMessageHandler(nil)
        hostRestore = nil
        hostSync?.setMessageHandler(nil)
        hostSync = nil
        hostDispatchEvent?.setMessageHandler(nil)
        hostDispatchEvent = nil
        hostRemoveMaskView?.setMessageHandler(nil)
        hostRemoveMaskView = nil
        flutterCreate = nil
        flutterSwitchTop = nil
        flutterRestore = nil
        flutterDestroy = nil
        flutterPush = nil
        flutterReplace = nil
        flutterPop = nil
        flutterMaybePop = nil
        flutterRemove = nil
        flutterNotifyPageVisible = nil
        flutterNotifyPageInvisible = nil
        flutterNotifyEnterForeground = nil
        flutterNotifyEnterBackground = nil
        flutterDispatchEvent = nil
        flutterCheckStyle = nil
        engine?.viewController = nil
        engine?.destroyContext()
        engine = nil
    }
}
